### PR TITLE
Add domain_blocks API

### LINKED ISSUE #260: Implement Domain Blocks API
Mastodon's Domain Blocks API allows a user to manage his list of blocked domains. The goal of this task is to implement the following endpoints of this API in BigBone:

- Get domain blocks
- Block a domain
- Unblock a domain

The Domain Blocks API as it is documented by Mastodon can be found here:
https://docs.joinmastodon.org/methods/domain_blocks/

Tip: A lot of other endpoints similar to this have already been implemented. In order to get an idea how to implement this, I suggest that you study the code upfront. Start with one of the `*Methods.kt` files.

Please consult the following documents to make sure your contribution meets our projects criteria:

- [Code of Conduct](https://github.com/andregasser/bigbone/blob/master/CODE_OF_CONDUCT.md)
- [Code Contribution Guidelines](https://github.com/andregasser/bigbone/blob/master/CONTRIBUTING.md)

### PR DESCRIPTION
**Pull request by PattaFeuFeu**
# Description

This PR implements the current state of the [domain_blocks API](https://docs.joinmastodon.org/methods/domain_blocks/) so that consumers of this library can get blocked domains for a user, block and unblock domains.

I have noticed that we use unnecessarily explicit code in the RxJava methods. With this PR, I have started to use the more concise `Single.fromCallable` and `Completable.fromAction` which automatically do the same as we previously implemented with `try { it.onSuccess(…) } catch (e: Throwable) { it.onError(e) }`.

I will soon create a PR to refactor all other Rx classes in that manner. This will also make our `onErrorIfNotDisposed` obsolete as the Rx versions I mentioned above already take care of dispose state. **Edit:** Created: https://github.com/andregasser/bigbone/pull/288

Closes #260.

# Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
    https://github.com/andregasser/bigbone/wiki/Mastodon-API-Coverage#methods needs to be updated after this has been merged!
- [ ] Dependency update / replacement

# How Has This Been Tested?

`gradle check`

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the documentation if necessary
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: bigbone/src/test/kotlin/social/bigbone/testtool/MockClient.kt [hunk 1]
```diff
@@ -94,15 +98,19 @@ object MockClient {
...
-        every { clientMock.post(ofType<String>(), any(), any()) } returns response
-        every { clientMock.postRequestBody(ofType<String>(), any()) } returns response
-        every { clientMock.put(ofType<String>(), any()) } returns response
-        every { clientMock.performAction(ofType<String>(), any()) } throws BigBoneRequestException("mock")
+        every { clientMock.delete(any<String>(), any<Parameters>()) } returns response
+        every { clientMock.get(any<String>(), any<Parameters>()) } returns response
+        every { clientMock.patch(any<String>(), any<Parameters>()) } returns response
+        every { clientMock.post(any<String>(), any<Parameters>(), any<Boolean>()) } returns response
+        every { clientMock.postRequestBody(any<String>(), any<RequestBody>()) } returns response
+        every { clientMock.put(any<String>(), any<Parameters>()) } returns response
```
Comment (PattaFeuFeu) [inline bigbone/src/test/kotlin/social/bigbone/testtool/MockClient.kt:106]: I’ve replaced this to better handle nullability of some of the variables. Using `any<T>()` ensures matches for null and non-null of type T.

File: bigbone/src/main/kotlin/social/bigbone/api/method/DomainBlocksMethods.kt [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    @JvmOverloads
+    @Throws(IllegalArgumentException::class)
+    fun getDomainBlocks(
+        range: Range = Range()
+    ): MastodonRequest<Pageable<String>> {
+        if (range.limit != null && range.limit > QUERY_RESULT_LIMIT) {
+            throw IllegalArgumentException(
+                "limit defined in Range must not be higher than $QUERY_RESULT_LIMIT but was ${range.limit}"
+            )
+        }
```
Comment (PattaFeuFeu) [inline bigbone/src/main/kotlin/social/bigbone/api/method/DomainBlocksMethods.kt:34]: Adds client-side validation per https://github.com/andregasser/bigbone/wiki/API-Support#error-handling

File: bigbone-rx/src/main/kotlin/social/bigbone/rx/RxDomainBlocksMethods.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    private val domainBlocksMethods = DomainBlocksMethods(client)
+
+    /**
+     * View domains the user has blocked.
+     * @param range optional Range for the pageable return value
+     * @see <a href="https://docs.joinmastodon.org/methods/domain_blocks/#get">Mastodon API documentation: methods/domain_blocks/#get</a>
+     * @return
+     */
+    fun getDomainBlocks(range: Range = Range()): Single<Pageable<String>> =
```
Comment (andregasser) [inline bigbone-rx/src/main/kotlin/social/bigbone/rx/RxDomainBlocksMethods.kt:25]: Please add `@JvmOverloads` annotation

File: bigbone/src/main/kotlin/social/bigbone/api/method/DomainBlocksMethods.kt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+     *
+     * @param domain Domain to block
+     * @see <a href="https://docs.joinmastodon.org/methods/domain_blocks/#block">Mastodon API documentation: methods/domain_blocks/#block</a>
+     * @throws IllegalArgumentException if [domain] is blank or contains spaces
+     */
+    @Throws(IllegalArgumentException::class)
+    fun blockDomain(
+        domain: String
+    ) {
+        require(domain.isNotBlank()) { "domain must not be blank" }
```
Comment (andregasser) [inline bigbone/src/main/kotlin/social/bigbone/api/method/DomainBlocksMethods.kt:55]: @PattaFeuFeu I think we should go through all methods and implement these checks everywhere where needed. Let's create a separate issue for this. Any objections?

Comment (PattaFeuFeu) [reply bigbone/src/main/kotlin/social/bigbone/api/method/DomainBlocksMethods.kt:55]: Sounds like a good idea 👍 I’ll create one!

Comment (PattaFeuFeu) [reply bigbone/src/main/kotlin/social/bigbone/api/method/DomainBlocksMethods.kt:55]: -> #293

**Review by andregasser — APPROVED** [review]
All good!

**Pull request merged**